include/dynamatic/Dialect/Handshake/HandshakeTypes.td

```diff
@@ -206,4 +206,178 @@ def ChannelOrSimpleControl : TypeConstraint<
+  "the front operand of variadic and non-variadic operand should have the same data type", CPred<[{
+    ([&](const mlir::OperandRange &operands, const mlir::Value &op) {
+      // Retrive all types of operands in the variadic.
+      ValueTypeRange<OperandRange> types = operands.getTypes();
+
+      // If variadic is empty (no front operand), returns true
+      if (types.empty())
+        return true;
+
+      if (mlir::isa<ControlType>(types.front())) {
+        // Pass if non-variadic type is also ControlType
+        return mlir::isa<ControlType>(op.getType());
+      }
+
+      // Pass if non-variadic type has the same data type
+      return mlir::cast<ChannelType>(types.front()).getDataType() == mlir::cast<ChannelType>(op.getType()).getDataType();
+    })
+  }] # "($" # variadic # ", $" # nonvariadic # ")">
+>;
+
+/// Multi-entity constraint ensuring that the all operands of variadic and non-variadic
+/// operand have the same data types.
+/// Also applies to ControlTypes, as some operations treat them interchangeably.
+class AllDataTypesMatchWithVariadic<string variadic, string nonvariadic> : PredOpTrait<
+  "the operands inside the variadic " # name # " and " # nonvariadic # " should all have the same type",
+  And<[
+    // Ensuring the data types of variadic[0], ..., variadic[n] is the same
+    AllDataTypesMatchWithinVariadic<variadic>.predicate,
+    // Ensuring the data types of variadic[0] and nonvariadic is the same
+    DataTypesMatchWithVariadicFront<variadic, nonvariadic>.predicate
+  ]>
+>;
+
+/// Constraint to ensure an input/output is of SimpleControl/Channel type.
+class IsSimpleHandshake<string name> : PredOpTrait<
+  name # " should be of SimpleControl or SimpleChannel type",
+  CPred<"cast<::dynamatic::handshake::ExtraSignalsTypeInterface>($" # name #
+    ".getType()).getNumExtraSignals() == 0">
+>;
+
+/// Merging Relationship of Extra Signals: For all intraOperand and extra signal
+/// A, intraOperand has the extra signal A if and only if there exists an interOperand
+/// which has the extra signal A.
+/// In Merge-like Ops interOperands are always provided as variadic.
+class MergingExtraSignals<string variadicInterOperands, list<string> intraOperands> : PredOpTrait<
+  "the extra signals of intraOperands and interOperands should satisfy \\\"Merging Relationship of Extra Signals\\\"",
+  CPred<[{
+    ([&](const mlir::OperandRange &interOperands, llvm::ArrayRef<Value> intraOperands) {
+
+      // Map to collect extra signals from interOperands
+      llvm::DenseMap<StringRef, const ExtraSignal *> extraSignals;
+
+      // Iterate over the type of each interOperand inside the variadic parameter
+      for (auto interOperandType : interOperands.getTypes()) {
+        for (const ExtraSignal &extraSignal :
+            cast<ExtraSignalsTypeInterface>(interOperandType).getExtraSignals()) {
+          if (extraSignals.contains(extraSignal.name)) {
+            if (*extraSignals.lookup(extraSignal.name) != extraSignal) {
```

Comment:
I thought the pointer comparison works here due to the uniqued property of MLIR, but it didn't.

```
extraSignals.lookup(extraSignal.name) != &extraSignal
```